honeybee/aperture.py

```diff
@@ -0,0 +1,506 @@
+            'Got {}'.format(type(value))
+        self._boundary_condition = value
+
+    @property
+    def parent(self):
+        """Parent Face if assigned. None if not assigned."""
+        return self._parent
+
+    @property
+    def has_parent(self):
+        """Boolean noting whether this Aperture has a parent Face."""
+        return self._parent is not None
+
+    def set_adjacency(self, other_aperture):
+        """Set this aperture to be adjacent to another.
+
+        Note that this method does not verify whether the other_aperture geometry is
+        co-planar or compatible with this one so it is recommended that a test
+        be performed before using this method in order to verify these criteria.
+        The Face3D.is_centered_adjacent() or the Face3D.is_geometrically_equivalent()
+        methods are both suitable for this purpose.
+
+        Args:
+            other_aperture: Another Aperture object to be set adjacent to this one.
+        """
+        assert isinstance(other_aperture, Aperture), \
+            'Expected Aperture. Got {}.'.format(type(other_aperture))
+        self.boundary_condition = Surface(other_aperture)
+        other_aperture.boundary_condition = Surface(self)
+
+    def overhang(self, depth, angle=0, indoor=False):
+        """Get a single overhang for this Aperture.
+
+        Args:
+            depth: A number for the overhang depth.
+            angle: A number for the for an angle to rotate the overhang in degrees.
+                Default is 0 for no rotation.
+            indoor: Boolean for whether the overhang should be generated facing the
+                opposite direction of the aperture normal (typically meaning
+                indoor geometry). Default: False.
+
+        Returns:
+            overhang: A Shade object for the Aperture overhang. Will be None if the
+                Aperture geometry does not have a flat top.
+        """
+        overhang = self.generate_louvers_by_number(
+            1, depth, angle=angle, indoor=indoor)
+        return overhang[0] if len(overhang) != 0 else None
+
+    def right_fin(self, depth, angle=0, indoor=False):
+        """Get a single vertical fin on the right side of this Aperture.
+
+        Args:
+            depth: A number for the fin depth.
+            angle: A number for the for an angle to rotate the fin in degrees.
+                Default is 0 for no rotation.
+            indoor: Boolean for whether the fin should be generated facing the
+                opposite direction of the aperture normal (typically meaning
+                indoor geometry). Default: False.
+
+        Returns:
+            fin: A Shade object for the Aperture fin. Will be None if the
+                Aperture geometry does not have a flat right side.
+        """
+        fin = self.generate_louvers_by_number(
+            1, depth, angle=angle, contour_vector=Vector3D(1, 0, 0), indoor=indoor)
+        return fin[0] if len(fin) != 0 else None
```

Comment:
I think for case of no fin / overhang, etc we should throw a warning so the user knows that no fin has been created. This will be more important when it comes to interfaces like Grasshopper and having a warning clarifies the non flat situation.

In case you are already throwing a warning in ladybug-geometry you can ignore this comment.

Reply:
I disagree that an exception is what you want to happen.  Let's say that you want to generate fins for a facade with 20 rectangular windows and 1 triangular window that sits under gabled roof.  You run this method over the apertures in the Face expecting to get fins for the 20 windows but instead you get an exception because of that pesky triangle in the roof.  You think that maybe you should remove the triangle aperture from the face and add it back afterwards but that is too much work.  So you wrap your whole code in a try/except loop to catch the `ValueError` raised by the one triangle aperture but this means that you don't properly catch all of the other ValueErrors that can happen while creating an overhang (like inputting a non-numerical string for the overhang depth instead of a number). So now you have some bad code that will cause errors.  If you instead read the docstring here and just added a `if overhang is None` in your code, things will be much better and much more resilient.
So, I would argue that, if there's any place to put an exception, it's on the plugin end and I honestly think that is overkill for a Grasshopper plugin because people can visually evaluate what the shade geometry looks like there.  Let them check the result and see if it is what they want rather than stooping them from visualizing the issue in the first place.

Reply:
I agree that this doesn't deserve an exception and I suggested a warning (warning.warns('No overhang is created.')). This will not raise an exception.

Reply:
Ah. I should have read more carefully.  Ok, I can import warnings and add a warning like what you say.

Reply:
I added the warnings in but I think that I underestimated how annoying they would be while running all of the tests.  You are going to have to show me how to turn the warnings off at some point. I can easily see myself getting to the point where I can't see the real errors amidst all of the warnings, especially if we are going to use them for cases like this one where I still a bit unsure that the warning is actually helpful.

Reply:
Would this work: https://stackoverflow.com/a/14463321/4394669